### PR TITLE
fix: various improvements to the base interface

### PR DESCRIPTION
**Pull request by davidjoy**
BREAKING CHANGE:
- App.authentication has been renamed App.authenticatedUser
- AuthenticationContext has been renamed to AppContext, and contains two subkeys: authenticatedUser and config.
- Applications are initialized via App.initialize instead of a stand-alone initialize function.

Other changes:
- The frontendAuthWrapper has been simplified and is nearly gone.
- App.initialize takes override functions for various parts of its startup sequence.
- App.initialize publishes events to add additional hooks into the startup process.

### CODE REVIEW

File: src/App.js [hunk 1]
```diff
@@ -1,26 +1,86 @@
...
 
 export const APP_TOPIC = 'APP';
+export const APP_BEFORE_INIT = `${APP_TOPIC}.BEFORE_INIT`;
+export const APP_CONFIGURED = `${APP_TOPIC}.CONFIGURED`;
+export const APP_VALIDATED = `${APP_TOPIC}.VALIDATED`;
+export const APP_AUTHENTICATED = `${APP_TOPIC}.AUTHENTICATED`;
+export const APP_I18N_CONFIGURED = `${APP_TOPIC}.I18N_CONFIGURED`;
+export const APP_LOGGING_CONFIGURED = `${APP_TOPIC}.LOGGING_CONFIGURED`;
+export const APP_ANALYTICS_CONFIGURED = `${APP_TOPIC}.ANALYTICS_CONFIGURED`;
+export const APP_BEFORE_READY = `${APP_TOPIC}.BEFORE_READY`;
```
Comment (abutterworth) [inline src/App.js]: Is the APP_ prefix needed?

File: src/App.js [hunk 2]
```diff
@@ -1,26 +1,86 @@
...
+      this.loggingService = loggingService;
+      this.custom = custom;
+
+      // Configuration
+      await this.override(handlers.configuration, overrideHandlers.configuration);
+      PubSub.publish(APP_CONFIGURED);
+
+      // Configuration validation
+      await this.override(handlers.validation, overrideHandlers.validation);
+      PubSub.publish(APP_VALIDATED);
```
Comment (abutterworth) [inline src/App.js]: I would expect CONFIGURED to indicate that the config was successfully applied and there for validated

File: src/App.js [hunk 3]
```diff
@@ -1,26 +1,86 @@
...
+
+      // Authentication
+      await this.override(handlers.authentication, overrideHandlers.authentication);
+      PubSub.publish(APP_AUTHENTICATED);
+
+      // Internationalization
+      await this.override(handlers.internationalization, overrideHandlers.internationalization);
```
Comment (abutterworth) [inline src/App.js]: Why use the full word here and not in the published event?

Comment (abutterworth) [reply src/App.js]: I think i18n is good here too unless you think it would generate confusion somewhere

File: src/App.js [hunk 4]
```diff
@@ -68,11 +120,19 @@ export default class App {
     return pick(this.config, keys);
   }
 
+  static async override(defaultHandler, overrideHandler) {
+    if (overrideHandler !== undefined) {
+      await overrideHandler(this);
```
Comment (abutterworth) [inline src/App.js]: Just a note that in the docs we should be clear that override handlers must return promises

File: src/AppContext.jsx [hunk 5]
```diff
@@ -0,0 +1,9 @@
+import React from 'react';
+import { defaultAuthenticatedUser } from './frontendAuthWrapper';
+
+const AppContext = React.createContext({
+  authenticatedUser: defaultAuthenticatedUser,
```
Comment (abutterworth) [inline src/AppContext.jsx]: Any downside to just naming this `user`?

File: src/AppProvider.jsx [hunk 6]
```diff
@@ -6,11 +6,11 @@ import { Router } from 'react-router-dom';
 
 import App from './App';
 import ErrorBoundary from './ErrorBoundary';
-import AuthenticationContext from './AuthenticationContext';
+import AppContext from './AppContext';
 
 const AppProvider = ({ store, children }) => (
```
Comment (abutterworth) [inline src/AppProvider.jsx]: I'm good with the name _I think_, but this isn't strictly a provider is it?

File: src/handlers/logging.js [hunk 7]
```diff
@@ -0,0 +1,5 @@
+import { configureLoggingService } from '@edx/frontend-logging';
+
+export default async function logging(app) {
+  configureLoggingService(app.loggingService);
```
Comment (abutterworth) [inline src/handlers/logging.js]: I don't quite understand what's happening here, but I know it's not new. We have a logging service and then configuring our own with it?

File: src/handlers/logging.test.js [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+  configureLoggingService,
+} from '@edx/frontend-logging';
+
+import logging from './logging';
+
+jest.mock('@edx/frontend-logging', () => ({
+  configureLoggingService: jest.fn(),
+}));
+
+it('should configure logging with the logging servie to use', () => {
```
Comment (abutterworth) [inline src/handlers/logging.test.js]:
```suggestion
it('should configure logging with the logging service to use', () => {
```

**Review by abutterworth — APPROVED** [review]
LGTM! 🥇

File: src/App.js (hunk 1 above)
Comment (davidjoy) [reply src/App.js]: Eh, just figured it'd be good to keep them nice and disambiguated from anything else.  `READY` and `ERROR` would be pretty vague constants.

File: src/App.js (hunk 3 above)
Comment (davidjoy) [reply src/App.js]: Oh, good catch.  The inconsistency is a mistake.

File: src/App.js (hunk 2 above)
Comment (davidjoy) [reply src/App.js]: That's fair.  Maybe I'll just put those two together again.  I sorta felt like VALIDATED might be used for other stuff, but that's probably overkill.

File: src/App.js (hunk 4 above)
Comment (davidjoy) [reply src/App.js]: Technically they don't have to.  `await` will do the right thing, which is nice. 

From here: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await

"If the value of the expression following the await operator is not a Promise, it's converted to a resolved Promise."

File: src/AppContext.jsx (hunk 5 above)
Comment (davidjoy) [reply src/AppContext.jsx]: We use the prefix "authenticated" in frontend-auth, fwiw.  Mostly it was just to keep it consistent.  But... Hm.  🤔

File: src/AppProvider.jsx (hunk 6 above)
Comment (davidjoy) [reply src/AppProvider.jsx]: It's a rollup of a bunch of providers that provide things, so I don't see why it isn't one. :)

Comment (davidjoy) [reply src/AppProvider.jsx]: Not sure there's a formal definition of "provider" beyond something that makes stuff available to the components it wraps.

File: src/handlers/logging.js (hunk 7 above)
Comment (davidjoy) [reply src/handlers/logging.js]: We're choosing a logging service implementation for the app.  It's basically what makes it so you can import "logError" from frontend-logging without needing to import NewRelicLoggingService instead and call its logError method.

**Pull request merged**